server/server.go

```diff
@@ -879,3 +880,17 @@ func (s *Server) publishConnectEvent(c *client) {
 
 	s.sendEventNotification(subject, s.createConnectEvent(c))
 }
+
+// This publishes disconnect events for every non-internal client.
+func (s *Server) publishDisconnectEventForAllClients() {
+	s.mu.Lock()
+	extClients := make([]*client, len(s.clients))
```

Comment:
This is wrong. You need `0, len(s.clients)`, otherwise, your resulting array will have `len(s.clients)` nil pointers first, then the ones you are going to append to. If you use this syntax, then in the loop below you would have to assign directly to the array index, as in `extClients[i] = c`